junit-engine-api/src/main/java/org/junit/gen5/engine/AbstractTestDescriptor.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class AbstractTestDescriptor implements TestDescriptor {
+
+	private final String uniqueId;
+	private TestDescriptor parent;
+	private final Set<TestDescriptor> children = new HashSet<>();
+
+	protected AbstractTestDescriptor(String uniqueId) {
+		this.uniqueId = uniqueId;
+	}
+
+	@Override
+	public boolean equals(Object other) {
+		if (other == null)
+			return false;
+		if (this.getClass() != other.getClass())
+			return false;
+		TestDescriptor otherDescriptor = (TestDescriptor) other;
+		return this.getUniqueId().equals(otherDescriptor.getUniqueId());
+	}
+
+	@Override
+	public int hashCode() {
```

Comment:
I didn't want to write anything about this, but I just can't stop myself, sorry: Even if it complies to Java's equals/hashCode, I think we could use this.getUniqueId().hashCode() which should return a more useful hash. ;-) With 42, we wouldn't be able to use the Descriptors in a HashSet/HashMap/etc. I think the hash of the uniqueId is a reasonable fallback, if subclasses do not implement hashCode() their own.